proposals/unsafe-evolution.md

```diff
@@ -0,0 +1,308 @@
+expression has a memory safety state of safe. If that expression has no nested expressions, it has a memory safety state of safe. The following expressions always have a memory safety state of unsafe:
+
+* [Pointer indirections][pointer-indirection]
+* [Pointer member access][pointer-member-access]
+* [Pointer element access][pointer-element-access]
+* Function pointer invocation
+* Element access on a fixed-size buffer
+* `stackalloc` under the conditions defined [below](#stack-allocation)
+
+In addition to these expressions, expressions can also conditionally introduce a memory safety state of unsafe if they depend on any symbol that is marked as `unsafe`. For example, calling a method
+that is marked as `unsafe` will cause the _invocation_expression_ to have a memory safety state of unsafe, even if the receiver and all arguments have a memory safety state of safe.
+
+> [!NOTE]
+> This section probably needs expansion to formally declare the various expression types and symbols that can change the safety state of an expression.
+
+#### Pointer types
+
+As mentioned, pointers become no longer inherently unsafe. Any references to unsafe contexts in [§24.3][pointer-types-spec] are deleted. Pointer types exist in normal C# and do not require `unsafe`
+to bring them into existence. The type definitions should be worked into [§8.1](https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/types.md#81-general) and its following sections, as
+other types.
+
+Similarly, [pointer conversions][pointer-conversions] should be worked into [§10](https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/conversions.md#10-conversions), with references to
+`unsafe` contexts removed.
+
+Similarly, [pointer expressions][pointer-expressions], except for [pointer indirection][pointer-indirection], [pointer member access][pointer-member-access], and
+[pointer element access][pointer-element-access], should be worked into [§12](https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/expressions.md), with references to `unsafe` contexts
+removed. No semantics change about the meaning of these expressions; the only change is that they no longer require an `unsafe` context to use.
+
+For [pointer indirection][pointer-indirection], [pointer member access][pointer-member-access], and [pointer element access][pointer-element-access], these operators remain unsafe, as these
+access memory that is not managed the runtime. They remain in [§24][unsafe-code.md], and continue to require an `unsafe` context to be used. Any use outside of an `unsafe` context is a warning.
+No semantics about these operators change; they still continue to mean exactly the same thing that they do today. These expressions always have a memory safety state of unsafe.
+
+The [fixed statement][fixed-statement] moves to [§13](https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/statements.md), with references to `unsafe` contexts removed.
```

Comment:
Probably need to make some cases `unsafe` still I think (relating to definite assignment)? E.g.
```csharp
static void SkipInit<T>(out T value)
{
    fixed (void* ptr = &value) ; // just take reference to mark it as assigned - should count as unsafe when for out variable?
}
```

A similar concern also exists with uninitialised variables:
```csharp
int i;
_ = &i; // no longer unsafe, but should be since i is uninitialised but now considered initialised?
i++;
```

In particular, maybe the syntax `&x` should be unsafe if `x` is `out` or not fully initialised? Or not counting towards definite assignment outside of `unsafe` (but this feels more confusing to me tbh).

Reply:
3741258f